### PR TITLE
Fetch group by attribute

### LINKED ISSUE #12964: Enable searching for Groups by their attributes in keycloak-spi
### Description

It is not currently possible to search for groups via attributes. This functionality was added for clients in these commits:
https://github.com/keycloak/keycloak/pull/7870
https://github.com/keycloak/keycloak/pull/8099
https://github.com/keycloak/keycloak/pull/8341

The goal would be to enable the same behavior for groups, allowing for SPI implementers to take advantage of this feature to enable more interactions with groups in extensions. 

### Discussion

https://github.com/keycloak/keycloak/discussions/12997

### Motivation

This feature would allow for SPI implementers to work with groups and group attributes in a dynamic way. For instance, setting a custom attribute that dictates a group's color would allow for an SPI implementer to create a service that enables organization admins to find all groups with a specific color. 

A real world use case would be internal use at sso.redhat.com, the goal is to utilize groups in order to make an association between an email domain, domain, and third party IdP. This would allow for enablement of third-party IdP auto-provisioning as well as allowing for users to be added to an sso.redhat.com organization without and manual intervention. 

As this is currently, this would require querying every group and then filtering out by the desired attribute; this is not scalable for a large amount of groups. 

### Details

A draft fork of the proposed enhancement exists here: https://github.com/agandhew/keycloak/tree/fetch-group-by-attribute

It's only in a stubbed state at the moment in order to start conversation around the topic

### PR DESCRIPTION
**Pull request by agandhew**
This PR will add a new feature to keycloak groups API to search groups based on groups attributes. This functionality is similar to what we have for clients.

closes [#12964](https://github.com/keycloak/keycloak/issues/12964)

### CODE REVIEW

**Comment by agandhew** [comment]
> @agandhew @Redhat-Alice would you be able to add a basic test at the storage provider level?

@alechenninger we have a separate branch where we have all the test cases related to this change. @RedHat-Alice Please cherry-pick your changes in this branch

File: integration/admin-client/src/main/java/org/keycloak/admin/client/resource/GroupsResource.java [hunk 1]
```diff
@@ -124,4 +124,7 @@ List<GroupRepresentation> groups(@QueryParam("search") String search,
     @Path("{id}")
     GroupResource group(@PathParam("id") String id);
 
+    @GET
+    @Produces(MediaType.APPLICATION_JSON)
+    List<GroupRepresentation> query(@QueryParam("q") String searchQuery);
```
Comment (alechenninger) [inline integration/admin-client/src/main/java/org/keycloak/admin/client/resource/GroupsResource.java:129]: Strictly speaking, we only need this feature in the SPI, so if it helps to reduce scope we can drop HTTP API from the PR. Needs maintainer feedback.

File: model/jpa/src/main/java/org/keycloak/models/jpa/JpaClientProviderFactory.java [hunk 2]
```diff
@@ -70,7 +70,7 @@ public String getId() {
     @Override
     public ClientProvider create(KeycloakSession session) {
         EntityManager em = session.getProvider(JpaConnectionProvider.class).getEntityManager();
-        return new JpaRealmProvider(session, em, clientSearchableAttributes);
+        return new JpaRealmProvider(session, em, clientSearchableAttributes,null);
```
Comment (alechenninger) [inline model/jpa/src/main/java/org/keycloak/models/jpa/JpaClientProviderFactory.java:73]: nit: no space after comma

File: model/jpa/src/main/java/org/keycloak/models/jpa/JpaGroupProviderFactory.java [hunk 3]
```diff
@@ -47,7 +65,7 @@ public String getId() {
     @Override
     public GroupProvider create(KeycloakSession session) {
         EntityManager em = session.getProvider(JpaConnectionProvider.class).getEntityManager();
-        return new JpaRealmProvider(session, em, null);
+        return new JpaRealmProvider(session, em, null,groupSearchableAttributes);
```
Comment (alechenninger) [inline model/jpa/src/main/java/org/keycloak/models/jpa/JpaGroupProviderFactory.java:68]: nit: no space after comma

File: model/jpa/src/main/resources/META-INF/jpa-changelog-18.0.1.xml [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    <changeSet author="keycloak" id="18.0.1-12964-supported-dbs">
+        <preConditions onSqlOutput="TEST" onFail="MARK_RAN">
+            <or>
+                <dbms type="mysql"/>
+                <dbms type="mariadb"/>
+                <dbms type="postgresql"/>
+                <dbms type="oracle"/>
+            </or>
+        </preConditions>
+        <createIndex tableName="GROUP_ATTRIBUTE" indexName="IDX_GROUP_ATT_BY_NAME_VALUE">
```
Comment (alechenninger) [inline model/jpa/src/main/resources/META-INF/jpa-changelog-18.0.1.xml:29]: Was this added for client attribute search also?

I wonder if this should be up to discretion of the Keycloak operator. If you use a lot of group attributes, but don't care to query on them, the index may be too costly of a change to introduce.

Even for our use case, if we could, I'd rather use a sparse index and only index the attributes we cared about querying rather than everything.

But again, if client attributes also have an index, then it probably makes sense to be consistent. On the other hand, client attributes are probably less common than group attributes since they cannot be managed through the admin console, unlike group attributes.

Comment (agandhew) [reply model/jpa/src/main/resources/META-INF/jpa-changelog-18.0.1.xml]: Client attributes also has index. We added this index to be consistent with client attribute table

File: integration/admin-client/src/main/java/org/keycloak/admin/client/resource/GroupsResource.java (hunk 1 above)
Comment (agandhew) [reply integration/admin-client/src/main/java/org/keycloak/admin/client/resource/GroupsResource.java:129]: As we referred the "search client by attributes" PRs, we belive there shouldn't be any major review comments as part of this. If we have to do any major refactoring in API side, we will drop API changes

File: model/legacy-private/src/main/java/org/keycloak/storage/GroupStorageManager.java [hunk 5]
```diff
@@ -71,6 +72,28 @@ public Stream<GroupModel> searchForGroupByNameStream(RealmModel realm, String se
         return Stream.concat(local, ext);
     }
 
+    /**
+     * Returns the group hierarchy with the given attribute name and attribute value for the given realm.
```
Comment (mhajas) [inline model/legacy-private/src/main/java/org/keycloak/storage/GroupStorageManager.java:76]: Should we have this Javadoc here? It seems unnecessary since we have it in the interface.

File: server-spi/src/main/java/org/keycloak/storage/group/GroupLookupProvider.java [hunk 6]
```diff
@@ -64,5 +65,19 @@ default List<GroupModel> searchForGroupByName(RealmModel realm, String search, I
...
+     * This is done until the group node does not have a parent (root group)
+     *
+     * @param realm Realm.
+     * @param attributes Case sensitive attributes.
+     * @param firstResult First result to return. Ignored if negative or {@code null}.
+     * @param maxResults Maximum number of results to return. Ignored if negative or {@code null}.
+     * @return Stream of root groups that have the given string in their name themself or a group in their child-collection has.
+     * The returned hierarchy contains siblings that do not necessarily have a matching name. Never returns {@code null}.
+     */
+    Stream<GroupModel> searchGroupsByAttributes(RealmModel realm, Map<String, String> attributes, Integer firstResult, Integer maxResults);
```
Comment (mhajas) [inline server-spi/src/main/java/org/keycloak/storage/group/GroupLookupProvider.java:81]: Do we need to have this in the `GroupLookupProvider`? What about `GroupProvider`? If I am not mistaken, eventually, we aim to merge all "special" interfaces into one. The reason being with the new store it doesn't make sense to have the methods split anymore.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/admin/group/GroupSearchTest.java [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+      // "filtered" attribute won't take effect when JPA is used
+      String[] expectedRes = isJpaStore() ? new String[]{GROUP1, GROUP2} : new String[]{GROUP2};
+      search(String.format("%s:%s %s:%s", ATTR_URL_NAME, ATTR_URL_VAL, ATTR_FILTERED_NAME, ATTR_FILTERED_VAL), expectedRes);
+    }
+    finally {
+      resetSearchableAttributes();
+    }
+  }
+
+//  @Test
```
Comment (mhajas) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/admin/group/GroupSearchTest.java:125]: Why is this commented?

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/admin/group/GroupTest.java [hunk 8]
```diff
@@ -160,27 +160,6 @@ public void testClientRemoveWithClientRoleGroupMapping() throws Exception {
         assertAdminEvents.assertEvent(testRealmId, OperationType.DELETE, AdminEventPaths.clientResourcePath(clientUuid), ResourceType.CLIENT);
     }
 
-    private GroupRepresentation createGroup(RealmResource realm, GroupRepresentation group) {
```
Comment (mhajas) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/admin/group/GroupTest.java:163]: Why is this removed? Is this somehow relevant to this PR?

File: server-spi/src/main/java/org/keycloak/models/RealmModel.java [hunk 9]
```diff
@@ -862,6 +862,15 @@ default List<GroupModel> searchForGroupByName(String search, Integer first, Inte
...
     Stream<GroupModel> searchForGroupByNameStream(String search, Integer first, Integer max);
 
+    /**
+     * Searches for groups by provided group attributes. Results that match the given filter are returned as a stream.
+     * @param attributes {@code String} Name of a group to be used as a filter.
+     * @param first {@code Integer} Index of the first desired group. Ignored if negative or {@code null}.
+     * @param max {@code Integer} Maximum number of returned groups. Ignored if negative or {@code null}.
+     * @return Stream of {@link GroupModel}. Never returns {@code null}.
+     */
+    Stream<GroupModel> searchGroupsByAttributes(Map<String, String> attributes, Integer first, Integer max);
```
Comment (mhajas) [inline server-spi/src/main/java/org/keycloak/models/RealmModel.java:872]: Do we need this method? I am aware, that we have quite a lot of similar methods in the `RealmModel` interface, however, we should aim for having only methods relevant to realms in `RealmModel` and avoid such shortcut methods in favour of `session.groups().****`.

**Review by mhajas — COMMENTED** [review]
Thanks for the PR @agandhew! See my comments in the code.

File: model/jpa/src/main/java/org/keycloak/models/jpa/JpaGroupProviderFactory.java (hunk 3 above)
Comment (agandhew) [reply model/jpa/src/main/java/org/keycloak/models/jpa/JpaGroupProviderFactory.java]: added space

File: model/jpa/src/main/java/org/keycloak/models/jpa/JpaClientProviderFactory.java (hunk 2 above)
Comment (agandhew) [reply model/jpa/src/main/java/org/keycloak/models/jpa/JpaClientProviderFactory.java]: added space

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/admin/group/GroupTest.java (hunk 8 above)
Comment (Redhat-Alice) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/admin/group/GroupTest.java:163]: I took the liberty of moving this to AbstractGroupTest to prevent code duplication

File: server-spi/src/main/java/org/keycloak/models/RealmModel.java (hunk 9 above)
Comment (Redhat-Alice) [reply server-spi/src/main/java/org/keycloak/models/RealmModel.java:872]: I think this was done to ease the process of collecting all of the attributes and checking that they are configured to be searchable before executing the query. Also to keep parity with `searchClientsByAttribute`

Comment (mhajas) [reply server-spi/src/main/java/org/keycloak/models/RealmModel.java:872]: Not sure what you mean by:

> ease the process of collecting all of the attributes and checking that they are configured to be searchable before executing the query

Can you be more specific? It seems that call:
```
realm.searchGroupsByAttributes(searchedAttributes, first, max)
```
is equivalent to
```
session.groups().searchGroupByAttributes(session, searchedAttributes, first, max)
```

The only difference is there is one more method call in a `RealmModel` implementation in the first case. My comment aims to get rid of this method call if it is possible.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/admin/group/GroupTest.java (hunk 8 above)
Comment (mhajas) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/admin/group/GroupTest.java:163]: Ah, I missed that. That makes sense

File: server-spi/src/main/java/org/keycloak/storage/group/GroupLookupProvider.java (hunk 6 above)
Comment (mhajas) [reply server-spi/src/main/java/org/keycloak/storage/group/GroupLookupProvider.java:81]: If you need this method here because you have your implementation of Group storage, feel free to leave this here.

File: model/map/src/main/java/org/keycloak/models/map/group/MapGroupProvider.java [hunk 10]
```diff
@@ -186,6 +187,12 @@ public Stream<GroupModel> searchForGroupByNameStream(RealmModel realm, String se
             }).sorted(GroupModel.COMPARE_BY_NAME).distinct();
     }
 
+    @Override
+    public Stream<GroupModel> searchGroupsByAttributes(RealmModel realm, Map<String, String> attributes, Integer firstResult, Integer maxResults) {
+        //TODO :Add implementation
+        return null;
```
[the remark this answers is not in the document]
Comment (hmlnarik) [reply model/map/src/main/java/org/keycloak/models/map/group/MapGroupProvider.java:193]: @Redhat-Alice Map implementation (including changes to existing storage implementations) are mandatory to be done here in this PR.

File: server-spi/src/main/java/org/keycloak/storage/group/GroupLookupProvider.java (hunk 6 above)
Comment (hmlnarik) [reply server-spi/src/main/java/org/keycloak/storage/group/GroupLookupProvider.java:81]: In the current interface state, this is correct placement.

File: model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java [hunk 11]
```diff
@@ -949,6 +953,39 @@ public Stream<GroupModel> searchForGroupByNameStream(RealmModel realm, String se
...
+        CriteriaBuilder builder = em.getCriteriaBuilder();
+        CriteriaQuery<GroupEntity> queryBuilder = builder.createQuery(GroupEntity.class);
+        Root<GroupEntity> root = queryBuilder.from(GroupEntity.class);
+
+        List<Predicate> predicates = new ArrayList<>();
+
+        predicates.add(builder.equal(root.get("realm"), realm.getId()));
+
+        for (Map.Entry<String, String> entry : filteredAttributes.entrySet()) {
+            String key = entry.getKey();
```
Comment (hmlnarik) [inline model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java:971]: `key` may be `null` or empty, in which case the entry should be ignored.

**Review by hmlnarik — COMMENTED** [review]
Thank you for the PR. I have left a couple of comments, please see inline.

File: model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java (hunk 11 above)
Comment (Redhat-Alice) [reply model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java:971]: I can't seem to reply to the map implementation comment directly so replying here, we do have a map implementation and something is up with git history. I'll make sure it gets added back into the commits thanks for calling it out

File: model/map/src/main/java/org/keycloak/models/map/group/MapGroupProvider.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (Redhat-Alice) [reply model/map/src/main/java/org/keycloak/models/map/group/MapGroupProvider.java:193]: we do have a map implementation and something is up with git history. I'll make sure it gets added back into the commits thanks for calling it out

[the remark this answers is not in the document]
Comment (Redhat-Alice) [reply model/map/src/main/java/org/keycloak/models/map/group/MapGroupProvider.java:193]: we do have a map implementation and something is up with git history. I'll make sure it gets added back into the commits thanks for calling it out

File: model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java [hunk 12]
```diff
@@ -61,14 +62,7 @@
...
 import org.keycloak.models.delegate.ClientModelLazyDelegate;
-import org.keycloak.models.jpa.entities.ClientAttributeEntity;
-import org.keycloak.models.jpa.entities.ClientEntity;
-import org.keycloak.models.jpa.entities.ClientScopeClientMappingEntity;
-import org.keycloak.models.jpa.entities.ClientScopeEntity;
-import org.keycloak.models.jpa.entities.GroupEntity;
-import org.keycloak.models.jpa.entities.RealmEntity;
-import org.keycloak.models.jpa.entities.RealmLocalizationTextsEntity;
-import org.keycloak.models.jpa.entities.RoleEntity;
+import org.keycloak.models.jpa.entities.*;
```
[the remark this answers is not in the document]
Comment (agandhew) [reply model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java]: Removed astrix

File: model/jpa/src/main/java/org/keycloak/models/jpa/entities/GroupEntity.java [hunk 13]
```diff
@@ -33,6 +33,7 @@
         @NamedQuery(name="getGroupIdsByNameContaining", query="select u.id from GroupEntity u where u.realm = :realm and u.name like concat('%',:search,'%') order by u.name ASC"),
         @NamedQuery(name="getGroupIdsByNameContainingFromIdList", query="select u.id from GroupEntity u where u.realm = :realm and lower(u.name) like lower(concat('%',:search,'%')) and u.id in :ids order by u.name ASC"),
         @NamedQuery(name="getGroupIdsFromIdList", query="select u.id from GroupEntity u where u.realm = :realm and u.id in :ids order by u.name ASC"),
+        @NamedQuery(name="getGroupIdsByAttributeNameAndAttributeValue", query="select distinct u.id from GroupEntity u join u.attributes a where u.realm = :realm and a.name = :name and a.value = :value"),
```
[the remark this answers is not in the document]
Comment (agandhew) [reply model/jpa/src/main/java/org/keycloak/models/jpa/entities/GroupEntity.java]: Removed unused query

File: model/legacy-private/src/main/java/org/keycloak/storage/GroupStorageManager.java (hunk 5 above)
Comment (agandhew) [reply model/legacy-private/src/main/java/org/keycloak/storage/GroupStorageManager.java]: Removed javadoc

File: server-spi/src/main/java/org/keycloak/models/RealmModel.java (hunk 9 above)
Comment (agandhew) [reply server-spi/src/main/java/org/keycloak/models/RealmModel.java:872]: We tried to align all the calls from `GroupsResource` in the same way that was implemented for other methods i.e. using `RealmModel` object. If you want, I can remove `searchGroupByAttributes` from `RealmModel` and all its implementation and replace it with keycloak session call that you shared.

File: model/map/src/main/java/org/keycloak/models/map/group/MapGroupProvider.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (agandhew) [reply model/map/src/main/java/org/keycloak/models/map/group/MapGroupProvider.java]: Added map implementation

File: server-spi/src/main/java/org/keycloak/models/RealmModel.java (hunk 9 above)
Comment (agandhew) [reply server-spi/src/main/java/org/keycloak/models/RealmModel.java]: Removed `searchGroupsByAttributes` from `RealmModel` and its implementations

File: testsuite/utils/src/main/java/org/keycloak/testsuite/util/cli/GroupCommands.java [hunk 14]
```diff
@@ -0,0 +1,59 @@
+package org.keycloak.testsuite.util.cli;
+
+import org.keycloak.models.GroupModel;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.RealmModel;
+
+public class GroupCommands {
+
+  public static class Create extends AbstractCommand {
```
[the remark this answers is not in the document]
Comment (agandhew) [reply testsuite/utils/src/main/java/org/keycloak/testsuite/util/cli/GroupCommands.java:9]: Replaced with groups

File: model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java (hunk 11 above)
Comment (agandhew) [reply model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java:971]: Added the logic to handle this case

**Comment by Redhat-Alice** [comment]
Something that I'm not confident we have covered is testing of all of the `Map` implementations for these changes. 

Reading through documentation and discussion based on Keycloak X I can't seem to find any concrete documentation on how to achieve this. Could we get some guidance for this effort?

**Review by mhajas — CHANGES_REQUESTED** [review]
> Something that I'm not confident we have covered is testing of all of the `Map` implementations for these changes.

Tests in the base testsuite (integration-arquillian) should be automatically running with the Map storage, unless they are somehow skipped for it. For example, `GroupSearchTest` should be running at the moment for all three map storages CHM, JPA, HotRod. It should be part of GHA. 

Thank you for the map storage addition @agandhew @Redhat-Alice! There are two things I am missing though. You added a new `SearchableField` named `ATTRIBUTE`, however, you haven't added support for this field in JPA and HotRod store.


### For JPA:
Something like this:
https://github.com/keycloak/keycloak/blob/c6c65ad10b7fa4df4c70abd81ac32dd31d5f8267/model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/client/JpaClientModelCriteriaBuilder.java#L77-L86

needs to be added here:
https://github.com/keycloak/keycloak/blob/main/model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/group/JpaGroupModelCriteriaBuilder.java#L74

Also, we probably need to add index similarly as in the legacy store here: https://github.com/keycloak/keycloak/tree/main/model/map-jpa/src/main/resources/META-INF/groups

### For HotRod
We need to make sure Ickle query can be created for `GroupModel.SearchableFields.ATTRIBUTE` by adding two lines similar to these:
https://github.com/keycloak/keycloak/blob/c9689252989d6d3fa8c1d20be3c5e84f8f7d2ba0/model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/IckleQueryMapModelCriteriaBuilder.java#L64
and 
https://github.com/keycloak/keycloak/blob/c9689252989d6d3fa8c1d20be3c5e84f8f7d2ba0/model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/IckleQueryWhereClauses.java#L62

And we need to add index for attributes for `HotRodGroupEntity`:
https://github.com/keycloak/keycloak/blob/main/model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/group/HotRodGroupEntity.java#L112-L113

similarly as it is done here:
https://github.com/keycloak/keycloak/blob/c9689252989d6d3fa8c1d20be3c5e84f8f7d2ba0/model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/client/HotRodClientEntity.java#L120-L122

**Comment by Redhat-Alice** [comment]
> For example, GroupSearchTest should be running at the moment for all three map storages CHM, JPA, HotRod. It should be part of GHA

Thanks for the feedback! Would you mind explaining what GHA is for clarity? I'm unfamiliar with the acronym

**Comment by mhajas** [comment]
@Redhat-Alice Sure, I am sorry. GHA means Github actions. These are some actions running with each opened PR, you can see the results in each PR at the bottom of the PR page. 

Unfortunately, GHA in this PR are not running automatically, because you are a first-time contributor to this repo therefore it is necessary that some maintainer allows it. Once you address my last comments, we will find someone who will allow it to run (currently JPA and HotRod tests would most probably fail).

**Comment by Redhat-Alice** [comment]
> Also, we probably need to add index similarly as in the legacy store here:

Looking at the changelog for group map jpa it looks like the name value pair is already indexed. Am I missing anything there?

**Comment by mhajas** [comment]
> Looking at the changelog for group map jpa it looks like the name value pair is already indexed. Am I missing anything there?

@Redhat-Alice I missed that. I didn't expect the attributes are indexed already. Then no changes are necessary for the changelog files. 

Could you please have a look at the results of GHA? There are some failures in `GroupSearchTest`.

**Comment by Redhat-Alice** [comment]
> Could you please have a look at the results of GHA? There are some failures in GroupSearchTest

Looking at the failures it appears that the error is "Realm Not Found" 

Stepping through the code locally all of the setup for importing the realm for each test appears to be working correctly and tests pass which is leading me to believe it's an issue with either the GHA or the test being set up incorrectly for GHA. 

Could you provide some feedback since the problem isn't reproducible locally?

**Comment by Redhat-Alice** [comment]
> Bumping this since I posted it before a holiday

**Comment by Redhat-Alice** [comment]
I have moved over group creation to utilize creators in an effort to prevent the realm issues that were being seen. The realm is no longer modified between tests so I suspect this should fix things

**Comment by Redhat-Alice** [comment]
Moved all new tests to utilize assertThat via hamcrest

File: model/jpa/src/main/resources/META-INF/jpa-changelog-20.0.0.xml [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+  ~ *
+  ~ * Unless required by applicable law or agreed to in writing, software
+  ~ * distributed under the License is distributed on an "AS IS" BASIS,
+  ~ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  ~ * See the License for the specific language governing permissions and
+  ~ * limitations under the License.
+  -->
+<databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.1.xsd">
+
+    <changeSet author="keycloak" id="18.0.1-12964-supported-dbs">
```
Comment (stianst) [inline model/jpa/src/main/resources/META-INF/jpa-changelog-20.0.0.xml:20]: Shouldn't these be `20`:
```suggestion
    <changeSet author="keycloak" id="20.0.0-12964-supported-dbs">
```

File: model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/group/HotRodGroupEntity.java [hunk 16]
```diff
@@ -109,6 +109,7 @@ public void setName(String name) {
     @ProtoField(number = 6)
     public String parentId;
 
+    @ProtoDoc("@Field(index = Index.YES, store = Store.YES)")
     @ProtoField(number = 7)
     public Set<HotRodAttributeEntityNonIndexed> attributes;
```
Comment (mhajas) [inline model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/group/HotRodGroupEntity.java:114]: We need to also change the type of the attribute.

```suggestion
    public Set<HotRodAttributeEntity> attributes;
```

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/group/JpaGroupModelCriteriaBuilder.java [hunk 17]
```diff
@@ -71,6 +77,16 @@ public JpaGroupModelCriteriaBuilder compare(SearchableModelField<? super GroupMo
                             cb.function("->", JsonbType.class, root.get("metadata"), cb.literal("fGrantedRoles")),
                             cb.literal(convertToJson(value[0]))))
                     );
+                } else if (modelField == GroupModel.SearchableFields.ATTRIBUTE) {
+                    validateValue(value, modelField, op, String.class, String.class);
+
+                    return new JpaGroupModelCriteriaBuilder((cb, query, root) -> {
+                        Join<JpaClientEntity, JpaClientAttributeEntity> join = root.join("attributes", JoinType.LEFT);
```
Comment (vramik) [inline model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/group/JpaGroupModelCriteriaBuilder.java:84]:
```suggestion
                        Join<JpaGroupEntity, JpaGroupAttributeEntity> join = root.join("attributes", JoinType.LEFT);
```

File: model/map/src/main/java/org/keycloak/models/map/group/MapGroupProvider.java [hunk 18]
```diff
@@ -186,6 +187,18 @@ public Stream<GroupModel> searchForGroupByNameStream(RealmModel realm, String se
...
 
+    @Override
+    public Stream<GroupModel> searchGroupsByAttributes(RealmModel realm, Map<String, String> attributes, Integer firstResult, Integer maxResults) {
+        DefaultModelCriteria<GroupModel> mcb = criteria();
+        mcb =  mcb.compare(GroupModel.SearchableFields.REALM_ID, Operator.EQ, realm.getId());
+        for (Map.Entry<String, String> entry : attributes.entrySet()) {
+            mcb = mcb.compare(GroupModel.SearchableFields.ATTRIBUTE, Operator.EQ, entry.getKey(), entry.getValue());
+        }
+
+        return tx.read(withCriteria(mcb).pagination(firstResult, maxResults, SearchableFields.ID ))
```
Comment (vramik) [inline model/map/src/main/java/org/keycloak/models/map/group/MapGroupProvider.java:198]: Why are the results ordered by `SearchableFields.ID`? Shouldn't it be rather ordered by `SearchableFields.NAME` ?

File: model/map/src/main/java/org/keycloak/models/map/storage/chm/MapFieldPredicates.java [hunk 19]
```diff
@@ -120,9 +120,11 @@ public class MapFieldPredicates {
         put(CLIENT_SCOPE_PREDICATES, ClientScopeModel.SearchableFields.NAME,      MapClientScopeEntity::getName);
 
         put(GROUP_PREDICATES, GroupModel.SearchableFields.REALM_ID,               MapGroupEntity::getRealmId);
+        put(GROUP_PREDICATES, GroupModel.SearchableFields.ID,                     MapGroupEntity::getId);
```
Comment (vramik) [inline model/map/src/main/java/org/keycloak/models/map/storage/chm/MapFieldPredicates.java:123]: This is not necessary, The `ID` is added already by calling `basePredicates(GroupModel.SearchableFields.ID);` at line 87.

File: testsuite/integration-arquillian/servers/auth-server/services/testsuite-providers/src/main/java/org/keycloak/testsuite/federation/HardcodedGroupStorageProvider.java [hunk 20]
```diff
@@ -67,6 +67,19 @@ public Stream<GroupModel> searchForGroupByNameStream(RealmModel realm, String se
...
 
+    @Override
+    public Stream<GroupModel> searchGroupsByAttributes(RealmModel realm, Map<String, String> attributes, Integer firstResult, Integer maxResults) {
+        if (Boolean.parseBoolean(component.getConfig().getFirst(HardcodedGroupStorageProviderFactory.DELAYED_SEARCH))) try {
+            Thread.sleep(5000l);
+        } catch (InterruptedException ex) {
+            Logger.getLogger(HardcodedGroupStorageProvider.class).warn(ex.getCause());
+            return Stream.empty();
+        }
+        //TODO : Incomplete implementation
```
Comment (vramik) [inline testsuite/integration-arquillian/servers/auth-server/services/testsuite-providers/src/main/java/org/keycloak/testsuite/federation/HardcodedGroupStorageProvider.java:78]: Is this TODO still valid? Is it planned to be addressed within this PR?

**Review by vramik — CHANGES_REQUESTED** [review]
Thank you @Redhat-Alice for the PR! I have few comments, please see inline!

**Review by mhajas — CHANGES_REQUESTED** [review]
Thanks for the changes. I added one more comment. Also, Github actions are failing. Could you please rebase?

File: model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java [hunk 21]
```diff
@@ -949,6 +953,42 @@ public Stream<GroupModel> searchForGroupByNameStream(RealmModel realm, String se
...
+
+            Join<GroupEntity, GroupAttributeEntity> attributeJoin = root.join("attributes");
+
+            Predicate attrNamePredicate = builder.equal(attributeJoin.get("name"), key);
+            Predicate attrValuePredicate = builder.equal(attributeJoin.get("value"), value);
+            predicates.add(builder.and(attrNamePredicate, attrValuePredicate));
+        }
+
+        Predicate finalPredicate = builder.and(predicates.toArray(new Predicate[0]));
+        queryBuilder.where(finalPredicate).orderBy(builder.asc(root.get("id")));
```
[the remark this answers is not in the document]
Comment (Redhat-Alice) [reply model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java]: Addressed this in latest push

File: model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/group/HotRodGroupEntity.java (hunk 16 above)
Comment (Redhat-Alice) [reply model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/group/HotRodGroupEntity.java]: Addressed this in latest push

File: model/map/src/main/java/org/keycloak/models/map/storage/chm/MapFieldPredicates.java (hunk 19 above)
Comment (Redhat-Alice) [reply model/map/src/main/java/org/keycloak/models/map/storage/chm/MapFieldPredicates.java]: removed

File: model/map/src/main/java/org/keycloak/models/map/group/MapGroupProvider.java (hunk 18 above)
Comment (Redhat-Alice) [reply model/map/src/main/java/org/keycloak/models/map/group/MapGroupProvider.java]: changed to name

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/group/JpaGroupModelCriteriaBuilder.java (hunk 17 above)
Comment (Redhat-Alice) [reply model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/group/JpaGroupModelCriteriaBuilder.java]: This should be fixed now

**Comment by miquelsi** [comment]
Internal test pipelines results are good and not introducing any regression:
https://keycloak-jenkins/job/universal-test-pipeline-server/2190/
https://keycloak-jenkins/job/universal-test-pipeline-adapters/1292/

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/group/JpaGroupModelCriteriaBuilder.java [hunk 22]
```diff
@@ -71,6 +75,16 @@ public JpaGroupModelCriteriaBuilder compare(SearchableModelField<? super GroupMo
                             cb.function("->", JsonbType.class, root.get("metadata"), cb.literal("fGrantedRoles")),
                             cb.literal(convertToJson(value[0]))))
                     );
+                } else if (modelField == GroupModel.SearchableFields.ATTRIBUTE) {
+                    validateValue(value, modelField, op, String.class, String.class);
+
+                    return new JpaGroupModelCriteriaBuilder((cb, query, root) -> {
+                        Join<JpaClientEntity, JpaGroupAttributeEntity> join = root.join("attributes", JoinType.LEFT);
```
Comment (vramik) [inline model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/group/JpaGroupModelCriteriaBuilder.java:82]:
```suggestion
                        Join<JpaGroupEntity, JpaGroupAttributeEntity> join = root.join("attributes", JoinType.LEFT);
```

**Review by vramik — CHANGES_REQUESTED** [review]

**Review by vramik — APPROVED** [review]

**Review by miquelsi — APPROVED** [review]
Approving from QE

**Review by mhajas — APPROVED** [review]
